design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+
+
+
+
+### Transaction State Store
+
+For FDB to perform a recovery, there is information that it needs to know about
+the database, such as the configuration, worker exclusions, backup status, etc.
+These values are stored into the database in the `\xff` system keyspace.
+However, during a recovery, FDB can't read this data from the storage servers,
+because recovery hasn't completed, so it doesn't know who the storage servers
+are yet.  Thus, a copy of this data is held in-memory on every proxy in the
+*transaction state store*, and durably persisted as a part of commits on the
+transaction logs.  Being durably stored on the transaction logs means the list
+of transaction logs can be fetched from the coordinators, and then used to load
+the rest of the information about the database.
+
+The in-memory storage engine writes an equal amount of mutations and snapshot
+data to a queue, an when a full snapshot of the data has been written, deletes
+the preceeding snapshot and begins writing a new one.  When backing an
+in-memory storage engine with the transaction logs, the
+`LogSystemDiskQueueAdapter` implements writing to a queue as committing
+mutations to the transaction logs with a special tag of `txsTag`, and deleting
+the preceeding snapshot as popping the transaction logs for the tag of `txsTag`
+until the version where the last full snapshot began.
+
+This means that unlike every other commit that is tagged and stored on the
+transaction logs, `txsTag` signifies data that is:
+
+1. Committed to infrequently
+2. Only peeked on recovery
+3. Popped infrequently, and a large portion of the data is popped at once
+4. A small total volume of data
+
+The most problematic of these is the infrequent popping.  Unpopped data will be
+spilled after some time, and if `txsTag` data is spilled and not popped, it
+will prevent the DiskQueue from being popped as well.  This will cause the
+DiskQueue to grow continuously.  The infrequent commits and small data volume
+means that there benefits of spill-by-reference over spill-by-value don't apply
+for this tag.
+
+Thus, even when configured to spill-by-reference, `txsTag` is spilled by value.
+
+### Disk Queue Recovery
+
+If a transaction log dies and restarts, all commits that were in memory at the
+time of the crash must be loaded back into memory.  Recovery is blocked on this
+process, as there might have been a commit to the transaction state store
+immediately before crashing, and that data needs to be fully readable during a
+recovery.
+
+In spill-by-value, the DiskQueue only ever contained commits that were also
+held in memory, and thus recovery would need to read up to 1.5GB of data.  With
+spill-by-reference, the DiskQueue could theoretically contain terrabytes of
+data. To keep recovery times boundedly low, FDB must still only read the
+commits that need to be loaded back into memory.
+
+This is done by persisting the location in the DiskQueue of the last spilled
```

Comment:
how does persisting the last spilled commit help you during recovery? I.E the last spilled commit may be at byte X but there could still be unpopped spilled commits that you also need to recover at X-1000 right?

Reply:
Ah ok I get it now. Once you perform a spill you're basically guaranteeing that there is no data in the DiskQueue before the location of the most recently spilled data in the DiskQueue that is required to recover in-memory data.

This would probably be more obvious if you copied and pasted the diagram of the DiskQueue you have above (the one with the line separating spilled data from "recent" data) and then had a few sentences explaining why that line separating the two is actually a hard boundary that you can rely upon

Reply:
I added an explicit sentence about the contract of spilling to explain that it provides a hard boundary.